python/pyspark/sql/session.py

```diff
@@ -414,6 +415,43 @@ def _createFromLocal(self, data, schema):
         data = [schema.toInternal(row) for row in data]
         return self._sc.parallelize(data), schema
 
+    def _createFromPandasWithArrow(self, df, schema):
+        """
+        Create a DataFrame from a given pandas.DataFrame by slicing the into partitions, converting
+        to Arrow data, then reading into the JVM to parallelsize. If a schema is passed in, the
+        data types will be used to coerce the data in Pandas to Arrow conversion.
+        """
+        import os
+        from tempfile import NamedTemporaryFile
+        from pyspark.serializers import ArrowSerializer
+        from pyspark.sql.types import from_arrow_schema, to_arrow_schema
+        import pyarrow as pa
+
+        # Slice the DataFrame into batches
+        step = -(-len(df) // self.sparkContext.defaultParallelism)  # round int up
+        df_slices = (df[start:start + step] for start in xrange(0, len(df), step))
+        arrow_schema = to_arrow_schema(schema) if schema is not None else None
+        batches = [pa.RecordBatch.from_pandas(df_slice, schema=arrow_schema, preserve_index=False)
+                   for df_slice in df_slices]
+
+        # write batches to temp file, read by JVM (borrowed from context.parallelize)
+        tempFile = NamedTemporaryFile(delete=False, dir=self._sc._temp_dir)
```

Comment:
This looks kind of duplicate with the main logic of `context.parallelize`. Maybe we can extract a common function from it.

Reply:
Yeah, it is - I didn't want to mess around with the `parallelize()` logic so I left it alone.  If we were to make a common function it would look like this

```python

def _dump_to_tempfile(data, serializer):
    tempFile = NamedTemporaryFile(delete=False, dir=self._temp_dir)
    try:
        serializer.dump_stream(c, tempFile)
        tempFile.close()
        readRDDFromFile = self._jvm.PythonRDD.readRDDFromFile
        return readRDDFromFile(self._jsc, tempFile.name, numSlices)
    finally:
        # readRDDFromFile eagerily reads the file so we can delete right after.
        os.unlink(tempFile.name)
```

and some changes to `parallelize` to call it
```python
# Make sure we distribute data evenly if it's smaller than self.batchSize
if "__len__" not in dir(c):
    c = list(c)    # Make it a list so we can compute its length
batchSize = max(1, min(len(c) // numSlices, self._batchSize or 1024))
serializer = BatchedSerializer(self._unbatched_serializer, batchSize)
jrdd = _dump_to_tempfile(c, serializer)
```

Let me know if you all think we should change this?

Reply:
I'd prefer less duplicate. Let's see if others support it.

Reply:
Sounds good to me.